### DISCUSSION TITLE
Create Link - list all pages (not files) and allow to select one

### DISCUSSION THREAD

**Discussion by zmila** [zim-desktop-wiki/zim-desktop-wiki#2562] [Ideas]
Hello. 

I wonder if there is possible to implement linking to an existing page in more simple way.

Say I select a word and press Ctrl+L, There is edit fields for link to, text, and button Browse. 
The button opens File browser and allows to find and select a file and insert link to file in form 
**../../Work/DB/Postgres.txt**

However for wiki the more natural way would be to list all existing pages and allow to select a page, 
and insert link like 
**Work:DB:Postgres** 

Is there a possibility to implement such functionality (adding new button near Browse)?

**Comment by sojusnik** [comment]
Have you tried to type something in the "link to" text field?

**Reply by zmila** [reply]
yes, I can type the whole path to desired page in the text field. 
in the case, if i remember where in hierarchy the target page is. 
otherwise a new page is created in wrong place.

**Reply by sojusnik** [reply]
You don't have to type the full path there, just type the name of the page you want to link and the full path to the page should appear. Press arrow down to select it and enter to accept, another enter to switch to the next text field and a final enter to link your text.

**Reply by zmila** [reply]
It would be nice. But when i just type the name of existing (in other tree branch) page, and hit enter.
Then new page is created as sibling of current. No full path is displayed.

Some lines from the debug log:

DEBUG: Store page: TagLibro:2024:01
DEBUG: Action: insert_link
DEBUG: Opening dialog "Edit Link"
DEBUG: Store page in background: TagLibro:2024:01
DEBUG: Dialog response OK
DEBUG: Closed dialog "Edit Link"
DEBUG: Activate link: Clojure
DEBUG: Saving page: TagLibro:2024:01
DEBUG: Store page: TagLibro:2024:01
INFO: Open page: TagLibro:2024:Clojure (TagLibro:2024:Clojure)

I see no way to select Work:Programming:Clojure, but to type the full path manually. Maybe my installation is wrong? 
I use Zim 0.75.2 on Ubuntu 22.

**Reply by sojusnik** [reply]
That's odd. Let's see what the coders here say.

**Reply by bttrx** [reply]
@zmila I think, Zim works as designed.

When you start typing the name of an existing page, Zim should show you a list of matching pages including full path. So either click on the desired page from the list or use the keyboard to navigate the list and then press `Enter`. (Works for me in Zim 0.75.2 on Ubuntu-based Linux Mint 20.2 MATE.)

**Reply by introt** [reply]
![image](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/26069787/2c3fa636-7f55-4d16-bcdc-e430d5d4d4d4)

You should be seeing a dropdown like as pictured here as you start typing the name of the page. If it is not appearing, either your page index or installation is messed up. Tools > Check and update index should fix the former, while the latter can unfortunately be rather hard to pin down.

**Reply by zmila** [reply]
@bttrx , thank you for your information. 

As I understand my installation is somehow broken :( 

Here are screenshots. I created new empty notebook. Created page ZimTest, and two sub-pages: BranchA and Branch B.
In page Aaa, sub-page of BranchA I type content: "link to page Bbb", select word Bbb, press Ctrl+L. 
The dialog popups with prefilled text "Bbb" (no path, no list of variants).
![CreatingLinkToBbb](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/1272885/72c0d44c-9ee9-4383-a1fb-61664d8ee454)

After press Enter, new page Bbb created under BranchA, as expected.
![CreatedLinkToBranchA_Bbb](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/1272885/87441d81-6af1-4666-bd58-dc860966abcb)

Having several pages with Bbb as part of name, I try to create link, typing "Bb" - no list of variants is displayed. 
![CreatingLink_NoListOfVariants](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/1272885/6e143af6-4124-4771-83c4-09a7124650b2)

Maybe, it's important, I'm on Gnome DE in Ubuntu. And thus some UI is not working correctly?

**Reply by zmila** [reply]
I searched some videos on Youtube about Zim. 
Now I see what you mean, when adding a link and typing text, a list with names appears allowing to select existing pages. 
My installation clearly has no such possibility. 

Also I checked, rearrange of pages hierarchy with mouse drag-n-drop is not working. 

I'll try to reinstall the application.

**Reply by zmila** [reply]
Reinstalled fresh version from PPA. 

Preparing to unpack .../zim_0.75.2-t202401041223~ubuntu22.04.1_all.deb ...
Unpacking zim (0.75.2-t202401041223~ubuntu22.04.1) ...
Setting up zim (0.75.2-t202401041223~ubuntu22.04.1) ...

The same behavior :(

**Reply by introt** [reply]
Do note that if you select existing text before opening the dialog, it will not offer you the menu before you enter or erase a character.

There have been similar issues related to (X)Wayland before (eg.  #746, #2444), so you might want to try whether the issue is present on both an X11 and Wayland sessions (if both are available on your system).

**Reply by introt** [reply]
> Also I checked, rearrange of pages hierarchy with mouse drag-n-drop is not working.

This has to be enabled in Preferences, it has been disabled by default due to being considered "wonky"

**Reply by zmila** [reply]
Ho! You're absolutely right! When I move Zim window to the left monitor, the autocomplete list in Create Link and Jump To works ok. 
So this is problem in GTK, not Zim. Thank you very much!

About DragNDrop - I see no such option in Preferences. Could you please show the location of it?

**Reply by introt** [reply]
Glad to be of assistance. You should find the DnD setting from here: https://github.com/zim-desktop-wiki/zim-desktop-wiki/issues/1699#issuecomment-986000790
